include/thread_pool/slotted_bag.hpp

```diff
@@ -0,0 +1,168 @@
+* The bag is divided into slots, each with a unique id, and each slot can be full or empty.
+* The bag supports filling and emptying slots, as well as the unordered emptying of any full slot. 
+* The bag supports multiple consumers, and a single producer per slot.
+*/
+template <template<typename> class Queue>
+class SlottedBag
+{
+    /**
+    * @brief The implementation of a single slot of the bag. Maintains state and id.
+    * @details State Machine:
+    *
+    *             +------ TryRemoveAny -------+-------------------------------+ 
+    *             |                           |                               | 
+    *             v                           |                               | 
+    *       +-----------+              +-------------+                 +---------------+           
+    *       | NotQueued | -- add() --> | QueuedValid | -- remove() --> | QueuedInvalid |
+    *       +-----------+              +-------------+                 +---------------+
+    *                                         ^                               |
+    *                                         |                               |
+    *                                         +------------- add() -----------+
+    */
+    struct Slot
+    {
+        enum class State
+        {
+            NotQueued,
+            QueuedValid,
+            QueuedInvalid,
+        };
+
+        std::atomic<State> state;
+        size_t id;
+
+        Slot()
+            : state(State::NotQueued)
+            , id(0)
+        {
+        }
+
+        Slot(const Slot&) = delete;
+        Slot& operator=(const Slot&) = delete;
+
+        Slot(Slot&& rhs) = default;
+        Slot& operator=(Slot&& rhs) = default;
+    };
+
+public:
+
+    /**
+    * @brief SlottedBag Constructor.
+    * @param size Power of 2 number - queue length.
+    * @throws std::invalid_argument if size is not a power of 2.
+    */
+    explicit SlottedBag(size_t size);
+
+    SlottedBag(SlottedBag& rhs) = delete;
+    SlottedBag& operator=(SlottedBag& rhs) = delete;
+
+    SlottedBag(SlottedBag&& rhs) = default;
+    SlottedBag& operator=(SlottedBag&& rhs) = default;
+
+    /**
+    * @brief fill Fill the specified slot.
+    * @param id The id of the slot to fill.
+    * @throws std::runtime_error if the slot is already full.
+    * @note Other exceptions may be thrown if the single-producer-per-slot
+    * semantics are violated.
+    */
+    void fill(size_t id);
+
+    /**
+    * @brief empty Empties the slot with the specified id.
+    * @param id The id of the slot to empty.
+    * @returns true if the state of the bag was changed, false otherwise. 
+    * @note The slot will be empty after the call regardless of the return value.
+    */
+    bool empty(size_t id);
+
+    /**
+    * @brief tryEmptyAny Try to empty any slot in the bag.
+    * @return a pair containing true upon success along with the id of the emptied slot, false otherwise with an id of UINT_MAX.
```

Comment:
UINT_MAX?  or ((size_t) -1)?  is there a constant for that?  numeric_limits<size_t>::max()?

Reply:
Fixed!